README.md
# Mute Bot Random Experiments

## Description

This project provide tools to experiment CRDT performances, in a collaboration.

In order to simulate a collaboration, a bot has been developed. It plays randomly an insert or a deletion in a document, with a probability to move in this. (cf. [mute-bot-random](###mute-bot-random))

---

## How to start an experiment

Requirements : you need to install **docker** before using this script

You can start an experiment with the start.sh script : `./start.sh ./path/docker-compose-file [number_of_experiment_simultaneous]`

Or by using docker stack directly :

```
mkdir './Results/myName'
export BOTRANDOM_EXPERIMENT_NAME=myName
docker stack deploy -c 'compose_file' experiment_stack_name
```

---

## Elements

The project is divided in four parts :

- _create-docker-compose_ : create a custom docker-compose file with your experiment's configuration
- _mute-bot-random_ : the bot source code
- _Results_ : some results done
- _results-exploit_ : tools in order to verify the snapshots integrity

### create-docker-compose

The configuration of the experiment is in `config.json`.
The JSON's structure is :

- _objective_: the total number of operation targeted
- _delay_: the time in ms before the bots will start the experiment
- _snapshot_: the number of operation between two snapshots
- _bots_ : the list of bots
  - _master_ : true if it's the master
  - _botname_ : The name of the bot. Take attention to have a unique name in the experiment
  - _time_ : the time between each operation
  - _operation_ : the number of operation the bot will operate
  - _deletion_ : the probability to have a delete operation in %
  - _move_ : the probability to move in the document before an operation in %

Usage : `node createDockerCompose.js`. The docker-compose file generated is created and placed in the docker-compose folder with the name `docker-compose-[timestamp].yml`

### mute-bot-random

You can build the project with `npm run build`

Usage: `npm start -- [options]`

```
Options:
  -V, --version                   output the version number
  -m, --master [url]              Master Url
  -p, --port [port]               The bot server port (default: 20001)
  -n, --namebot [name]            the name of the bot (default: "Bob")
  -o, --objective [nbOperations]  The number of operation (default: 10)
  --operation [nbOperations]      THe number of operation the bot will make (default: 10)
  --deletion [deletion]           The probability to have a deletion instead of an insertion (default: 0)
  --deplacement [deplacement]     The probability to move the cursor (default: 0)
  --time [ms]                     The time between each operations (default: 1000)
  --delay [ms]                    The time before starting (default: 5000)
  --address [adr]                 the adress of the node for exemple ws://[adr]:20001 (default: "localhost")
  -s, --snapshot [nbOperation]    save a snapshot of the structure every [nbOperation] operations (default: 10)
  -h, --help                      output usage information
```

### Results

Results are save in a folder with timestamp as name. There is two kind of file :

- Logs : all operation made and received by a node are in this file `Logs.[botname].json`
- Snapshot : The serialized sequence crdt given with a certain number operations `Snaphot.[nbOperation].[botname].json`

### result-exploit

You can build the project with `npm run build`

The first function is to compare str from snapshots in the `./input` folder.
You have to copy Snapshot file you want to compare in that folders, and only these.
You can run `npm run start`

The second, is experiment on an other operation apply on a specific snapshot. This measure time to apply an operation on the structure. You can run `npm start -- -e /snapshot/path`


mute-bot-random/CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.1"></a>

## [2.0.1](https://github.com/coast-team/mute-bot-random/compare/v1.0.0...v2.0.1) (2021-10-07)

### Bug Fixes

- **bot-random:** increase objective by nb of generated rename operations ([9460521](https://github.com/coast-team/mute-bot-random/commit/9460521))
- **bot-random:** log properly local rename operations ([f5466ef](https://github.com/coast-team/mute-bot-random/commit/f5466ef))
- **bot-random:** trigger rename op as soon as possible ([653b3af](https://github.com/coast-team/mute-bot-random/commit/653b3af))
- **consolelog:** change console's log with local and remote operations, add chmod in start.sh ([a94cb82](https://github.com/coast-team/mute-bot-random/commit/a94cb82))
- **create-docker-compose:** fix depends_on ([d61fe20](https://github.com/coast-team/mute-bot-random/commit/d61fe20))
- **createdockerfile:** fix the addresses of the ws server argument ([f2a13f6](https://github.com/coast-team/mute-bot-random/commit/f2a13f6))
- **package.json:** fix precommit hook ([f839db4](https://github.com/coast-team/mute-bot-random/commit/f839db4))
- **rls-experimentparts:** fix experiment ([8e54506](https://github.com/coast-team/mute-bot-random/commit/8e54506))
- **snapshot:** fix asynchronous problem ([1752f51](https://github.com/coast-team/mute-bot-random/commit/1752f51))

### Features

- **bot-random:** add option to choose interval between rename ops ([7852a3d](https://github.com/coast-team/mute-bot-random/commit/7852a3d))
- **bot-random:** add option to choose nb of bots issuing rename ops ([0edd558](https://github.com/coast-team/mute-bot-random/commit/0edd558))
- **bot-random:** enable concurrent rename ops ([e4d0ac5](https://github.com/coast-team/mute-bot-random/commit/e4d0ac5))
- **buffer:** buffer operations before writing logs in file ([321f021](https://github.com/coast-team/mute-bot-random/commit/321f021))
- **check-validity:** add check on the identifiers order ([955ace9](https://github.com/coast-team/mute-bot-random/commit/955ace9))
- **check-validity:** add lists of simulations ([71f3589](https://github.com/coast-team/mute-bot-random/commit/71f3589))
- **create-docker-compose:** simplify required config ([ff8b036](https://github.com/coast-team/mute-bot-random/commit/ff8b036))
- **docker-compose:** add docker-compose-rls-150k-10b ([5bd5ab3](https://github.com/coast-team/mute-bot-random/commit/5bd5ab3))
- **dockerfile:** now the service name is the botname choosen in the config file ([32c5a1d](https://github.com/coast-team/mute-bot-random/commit/32c5a1d))
- **dotted:** add dottedLogootsplit ([8180060](https://github.com/coast-team/mute-bot-random/commit/8180060))
- **exploit:** add configuration file for result-exploit ([81e72df](https://github.com/coast-team/mute-bot-random/commit/81e72df))
- **figures:** add scripts to generate plots of the snapshot sizes ([6d8a7d2](https://github.com/coast-team/mute-bot-random/commit/6d8a7d2))
- **figures:** generate plot of the integration times of local ops ([ffd60ce](https://github.com/coast-team/mute-bot-random/commit/ffd60ce))
- **figures:** generate plot of the integration times of remote ops ([8a4c59b](https://github.com/coast-team/mute-bot-random/commit/8a4c59b))
- **figures:** generate plot of the integration times of rename ops ([7d9b239](https://github.com/coast-team/mute-bot-random/commit/7d9b239))
- **generate-snapshots:** add script to generate intermediate snapshots ([e4ae6f2](https://github.com/coast-team/mute-bot-random/commit/e4ae6f2))
- **integration-times:** aggregate results into a single file ([26e27b8](https://github.com/coast-team/mute-bot-random/commit/26e27b8))
- **integration-times-insert:** merge scripts for LS/RLS & add new stats ([b0f0177](https://github.com/coast-team/mute-bot-random/commit/b0f0177))
- **integration-times-rename:** add new stats ([311cc1c](https://github.com/coast-team/mute-bot-random/commit/311cc1c))
- **logs:** display the number of operation integrated instead of only local ([566823d](https://github.com/coast-team/mute-bot-random/commit/566823d))
- **mute-bot-random:** add support for RenamableLogootSplit ([c636f78](https://github.com/coast-team/mute-bot-random/commit/c636f78))
- **mute-bot-random:** replace bufferCount with bufferTime ([49a6de3](https://github.com/coast-team/mute-bot-random/commit/49a6de3))
- **mute-bot-random:** wait 60 before exiting process once terminated ([686df31](https://github.com/coast-team/mute-bot-random/commit/686df31))
- **mute-bot-random:** write final content to file before exiting ([5cdd4d9](https://github.com/coast-team/mute-bot-random/commit/5cdd4d9))
- **mutecore:** update mute-core with the new experimental version ([8b3263c](https://github.com/coast-team/mute-bot-random/commit/8b3263c))
- **network:** fix neighbour's list bug, add reconnection attempt ([5936421](https://github.com/coast-team/mute-bot-random/commit/5936421))
- **network:** increase the number of try to connect bots each other ([d205d5f](https://github.com/coast-team/mute-bot-random/commit/d205d5f))
- **package:** update mute-core ([1dbdd8a](https://github.com/coast-team/mute-bot-random/commit/1dbdd8a))
- **package:** update mute-core and add chmod line in start.sh ([33649c9](https://github.com/coast-team/mute-bot-random/commit/33649c9))
- **param:** add two parametters for buffer and logsnumber ([81553e6](https://github.com/coast-team/mute-bot-random/commit/81553e6))
- **resultexploit:** add script to run ressults-exploit in a docker container ([26357a0](https://github.com/coast-team/mute-bot-random/commit/26357a0))
- **results-exploit:** add script to compute sizes of snapshots ([98af319](https://github.com/coast-team/mute-bot-random/commit/98af319))
- **results-exploit:** add script to replay a log ([b756b26](https://github.com/coast-team/mute-bot-random/commit/b756b26))

mute-bot-random/src/index.ts
import { Strategy } from '@coast-team/mute-core'
import * as program from 'commander'
import { BotRandom } from './BotRandom'

// Retrieve version from package.json
let version: string
try {
  version = require('../package.json').version
} catch (err) {
  try {
    version = require('./package.json').version
  } catch {
    log.error(err)
    version = ''
  }
}

program
  .version(version)
  .option('-m, --master [url]', 'the URL of the master bot')
  .option('-p, --port [port]', 'the bot server port', 20001)
  .option('-n, --namebot [name]', 'the name of the bot', 'Bot0')
  .option('-o, --objective [nbOperations]', 'the number of operations of the simulation', 10)
  .option('--nb-operations [nbOperations]', 'the number of operations the bot will perform', 10)
  .option(
    '--deletion [deletion]',
    'the probability to perform a deletion instead of an insertion',
    20
  )
  .option('--move [move]', 'the probability to move the cursor before each local operation', 0)
  .option(
    '--nb-renaming-bots [nbRenamingBots]',
    'the number of bots which will issue rename operations',
    1
  )
  .option(
    '--rename-op-interval [renameOpInterval]',
    'issue a rename op every [renameOpInterval] operation',
    10
  )
  .option('--operation-interval [ms]', 'the time between each operations', 1000)
  .option('--delay [ms]', 'the initial delay before starting the simulation', 5000)
  .option('--address [adr]', 'the address of the node, for example ws://[adr]:20001', 'localhost')
  .option('--buffer-size [nb]', 'save logs in file every [nb] operations', 10)
  .option('--log-interval [nb]', 'log every [nb] operation in the console', 100)
  .option(
    '-s, --snapshot [nbOperation]',
    'save a snapshot of the structure every [nbOperation] operations',
    10
  )
  .option('--strategy [strat]', 'The strategy to use', 'ls')
  .parse(process.argv)

console.log('Start : port', program.port, ' - master', program.master)

const port = parseInt(program.port, 10)
const objective = parseInt(program.objective, 10)
const snapshot = parseInt(program.snapshot, 10)
const bufferSize = parseInt(program.bufferSize, 10)
const logInterval = parseInt(program.logInterval, 10)
const nbOperations = parseInt(program.nbOperations, 10)
const operationInterval = parseInt(program.operationInterval, 10)
const pDeletion = parseInt(program.deletion, 10)
const pMove = parseInt(program.move, 10)
const nbRenamingBots = parseInt(program.nbRenamingBots, 10)
const renameOpInterval = parseInt(program.renameOpInterval, 10)
const delay = parseInt(program.delay, 10)

let strat = Strategy.LOGOOTSPLIT
switch (program.strategy) {
  case 'dls':
    strat = Strategy.DOTTEDLOGOOTSPLIT
    break
  case 'fifodls':
    strat = Strategy.FIFODOTTEDLOGOOTSPLIT
    break
  case 'rls':
    strat = Strategy.RENAMABLELOGOOTSPLIT
    break
  default:
    break
}

const bot = new BotRandom(
  program.namebot,
  program.master,
  port,
  program.address,
  objective,
  snapshot,
  strat,
  nbRenamingBots,
  renameOpInterval,
  bufferSize,
  logInterval
)

setTimeout(() => {
  bot.doChanges(nbOperations, operationInterval, pDeletion, pMove)
}, delay)
